File: AlphaApps/SmartHome-App-Series/SmartHome-Version-3.0/java/mode/wifi/DeviceViewHolder.java
package com.sanjeev.stephan.murmu.ctnetwork.smarthome.mode.wifi;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.Switch;
import android.widget.TextView;

import com.sanjeev.stephan.murmu.ctnetwork.smarthome.R;
import com.sanjeev.stephan.murmu.ctnetwork.smarthome.mode.wifi.database.DeviceDatabase;

public class DeviceViewHolder extends RecyclerView.ViewHolder {

    public ImageView deviceImage;
    public TextView deviceName,deviceLocation,deviceStatus;
    public Switch deviceState;
    public DeviceViewHolder(@NonNull View itemView, final DeviceAdapter.OnDeviceItemClickListener deviceListener)
    {
        super(itemView);

        deviceImage = itemView.findViewById(R.id.id_device_model_image);
        deviceName = itemView.findViewById(R.id.id_device_model_name);
        deviceLocation = itemView.findViewById(R.id.id_device_model_location);
        deviceStatus = itemView.findViewById(R.id.id_device_model_status);
        deviceState = itemView.findViewById(R.id.id_device_model_state);

        itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                if(deviceListener != null)
                {
                    int position = getAdapterPosition();

                    if(position != RecyclerView.NO_POSITION) //NO_POSITION means (-1)
                    {
                        deviceListener.onItemClickListener(position);
                    }
                }
            }
        });

        deviceState.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton compoundButton, boolean b) {

                if(deviceListener != null)
                {
                    int position = getAdapterPosition();

                    if(position != RecyclerView.NO_POSITION)
                    {
                        deviceListener.onStateChangeListener(position,b);
                    }
                }
            }
        });

    }


}

File: projects/smarthome/SmartHome-Version-Kotlin/app/src/main/java/com/sanjeev/stephan/murmu/smarthome/sample/DeviceViewHolder.kt
package com.sanjeev.stephan.murmu.smarthome.sample

import android.view.View
import android.widget.CompoundButton
import android.widget.ImageView
import android.widget.Switch
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.sanjeev.stephan.murmu.smarthome.R

class DeviceViewHolder(itemView: View, listener: DeviceAdapter.OnMyItemClickListener) :
    RecyclerView.ViewHolder(itemView) {
    var deviceName: TextView = itemView.findViewById(R.id.id_device_item_name)
    var deviceStatus: TextView = itemView.findViewById(R.id.id_device_item_status)
    var deviceTime: TextView = itemView.findViewById(R.id.id_device_item_status_data_time)
    var deviceImage: ImageView = itemView.findViewById(R.id.id_device_item_image)
    var deviceSwitch: Switch = itemView.findViewById(R.id.id_device_switch)

    init {

        itemView.setOnClickListener(View.OnClickListener {
            val position: Int = adapterPosition
            if (position != RecyclerView.NO_POSITION) listener.onRecyclerItemClickListener(position)
        })

        deviceName.setOnClickListener(View.OnClickListener {
            //_: View? ->

            val position: Int = adapterPosition
            if (position != RecyclerView.NO_POSITION) {
                listener.onRecyclerItemClickListener(position)
            }

        }) //deviceName onClick Block ends


        deviceImage.setOnClickListener(View.OnClickListener {
            val position: Int = adapterPosition
            if (position != RecyclerView.NO_POSITION) listener.onRecyclerImageClickListener(position)
        }) //device Image Block Ends

        deviceSwitch.setOnCheckedChangeListener(CompoundButton.OnCheckedChangeListener()
        { _: CompoundButton?, isChecked: Boolean ->

            val position = adapterPosition
            if (position != RecyclerView.NO_POSITION) {

                when(isChecked) {
                    true -> changeImage(R.drawable.bulb_on,"Turned ON")
                    false -> changeImage(R.drawable.bulb_off,"Turned OFF")
                }
            } })

    }
       private fun changeImage(drawableImage: Int,statusMsg : String)
        { deviceImage.setImageResource(drawableImage)
          deviceStatus.text = statusMsg
          deviceTime.text = "NOW"
        }
}